nova/metrics.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2023 SAP SE
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from enum import auto
+from enum import Enum
+import logging
+from typing import Optional
+from typing import Union
+
+import statsd
+
+import nova.conf
+
+LOG = logging.getLogger(__name__)
+CONF = nova.conf.CONF
+
+class MetricType(Enum):
+    """https://statsd.readthedocs.io/en/v4.0.1/types.html"""
+    COUNTER = auto()
+    TIMER = auto()
+    GAUGE = auto()
+    SET = auto()
+
+
+_METRICS_PREFIX = "openstack.compute"
+_STATSD_CLIENT: Optional[statsd.StatsClient] = None
+
+def start():
+    global _STATSD_CLIENT
+    _STATSD_CLIENT = statsd.StatsClient("localhost", CONF.statsd_port)
+
+def _send_metric(tag: str, value: Union[int, float, None],
+                 m_type: MetricType = MetricType.GAUGE):
+    if _STATSD_CLIENT is None:
+        return
```

Comment:
My preference would be to create the statsd client lazily in `_send_metric`.
The client is an sending udp-socket, so it should be rather quick. No connection will be opened.

The reason for that is that the whole metric part would be entirely contained to the call-site and not modification of `nova.service.Service`.

To split it up explicitly in start would make for me more sense, if the start either has a lot of latency, or a fair chance of failing  (i.e. tcp-connection), or it is desired to orchestrate it somehow. I.e. there needs to be an order of how things are executed, for example, you do not want the metrics be send before exactly the time you call it in `nova.service.Service`

Not a required change from my side, just my preference.

Reply:
```ipython
In [2]: from statsd import StatsClient
In [3]: %timeit StatsClient("localhost", 9134)
59.8 µs ± 1.06 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
sounds quick enough.